File: 01-week-1--starter-algorithms/03-day-4--selection-sort/ruby/selection_sort.rb
def selection_sort(arr)
  sorted = []
  arr.size.times do
    min = arr.min
    idx = arr.index min
    sorted << min
    arr.delete_at idx
  end
  sorted
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort([3, -1, 5, 2])

  puts

  long_input = []

  100.times { long_input << rand }

  start_time = Time.now

  1000.times do
    selection_sort([2, 1])
    selection_sort(long_input)
  end

  puts ((Time.now - start_time) / 2000)
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 01-week-1--starter-algorithms/03-day-4--selection-sort/javascript/selection_sort.js
function selectionSort(arr) {
  let sorted = []
  if (arr.length === 0) {
    return sorted
  }
  sorted.push(arr.shift())
  while (arr.length > 0) {
    let value = arr.shift()
    let idx = sorted.findIndex(element => element >= value)
    if (idx === -1) {
      sorted.push(value)
    } else {
      sorted.splice(idx, 0, value)
    }
  }
  return sorted
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSort([3, -1, 5, 2]));

  console.log("");

  // BENCHMARK HERE, and print the average runtime
  const longInput = [];

  for (let i = 0; i < 100; ++i) {
    longInput.push(Math.random());
  }
}

module.exports = selectionSort;

// Please add your pseudocode to this file
// And a written explanation of your solution
